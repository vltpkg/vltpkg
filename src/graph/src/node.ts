import {
  DepID,
  DepIDTuple,
  getId,
  hydrateTuple,
  splitDepID,
} from '@vltpkg/dep-id'
import { typeError } from '@vltpkg/error-cause'
import { Spec, SpecOptions } from '@vltpkg/spec'
import { Integrity, ManifestMinified } from '@vltpkg/types'
import { DependencyTypeShort } from './dependencies.js'
import { Edge } from './edge.js'
import { GraphLike, NodeLike } from './types.js'

export type NodeOptions = SpecOptions & {
  projectRoot: string
  graph: GraphLike
}

export class Node implements NodeLike {
  get [Symbol.toStringTag]() {
    return '@vltpkg/graph.Node'
  }

  #options: SpecOptions
  #location?: string

  #optional = false
  /**
   * True if a node is only reachable via optional or peerOptional edges from
   * any importer.
   *
   * Setting this to false, if previously set to true, will also unset
   * the flag on any optional-flagged non-optional dependencies.
   */
  get optional() {
    return this.#optional
  }
  set optional(optional: boolean) {
    const before = this.#optional
    this.#optional = optional
    if (before && !optional) {
      // unset for all deps, as well
      for (const { to, optional } of this.edgesOut.values()) {
        if (!optional && to?.optional) to.optional = false
      }
    }
  }

  isOptional(): this is Node & { optional: true } {
    return this.#optional
  }

  #dev = false
  /**
   * True if a node is only reachable via dev edges from any importer.
   *
   * Setting this to false, if previously set to true, will also unset
   * the flag on any dev-flagged non-dev dependencies.
   */
  get dev() {
    return this.#dev
  }
  set dev(dev: boolean) {
    const before = this.#dev
    this.#dev = dev
    if (before && !dev) {
      // unset for all deps, as well
      for (const { to, dev } of this.edgesOut.values()) {
        if (!dev && to?.dev) to.dev = false
      }
    }
  }

  isDev(): this is Node & { dev: true } {
    return this.#dev
  }

  /**
   * List of edges coming into this node.
   */
  edgesIn = new Set<Edge>()

  /**
   * List of edges from this node into other nodes. This usually represents
   * that the connected node is a direct dependency of this node.
   */
  edgesOut = new Map<string, Edge>()

  /**
   * A reference to the {@link @DepID} this node represents in the graph.
   */
  id: DepID

  /**
   * True if this node is an importer node.
   */
  importer = false

  /**
   * True if this node is the project root node.
   */
  mainImporter = false

  /**
   * A reference to the graph this node is a part of.
   */
  graph: GraphLike

  /**
   * The manifest integrity value.
   */
  integrity?: Integrity

  /**
   * The manifest this node represents in the graph.
   */
  manifest?: ManifestMinified

  /**
   * Project where this node resides
   */
  projectRoot: string

  /**
   * For registry nodes, this is the registry we fetched them from.
   * Needed because their un-prefixed dependencies need to come from
   * the same registry, if it's not the default.
   */
  registry?: string

  /**
   * The name of the package represented by this node, this is usually
   * equivalent to `manifest.name` but in a few ways it may differ such as
   * nodes loaded from a lockfile that lacks a loaded manifest.
   * This field should be used to retrieve package names instead.
   */
  #name?: string
  get name() {
    if (this.#name) return this.#name
    this.#name = this.id
    return this.#name
  }

  /**
   * The location of the node_modules folder where this node's edgesOut
   * should be linked into. For nodes in the store, this is the parent
   * directory, since they're extracted into a node_modules folder
   * side by side with links to their deps. For nodes outside of the store
   * (ie, importers and arbitrary link deps) this is the node_modules folder
   * directly inside the node's directory.
   */
  get nodeModules() {
    const loc = this.location
    return this.inVltStore() ?
        loc.substring(0, loc.length - this.name.length - 1)
      : loc + '/node_modules'
  }

  /**
   * The version of the package represented by this node, this is usually
   * equivalent to `manifest.version` but in a few ways it may differ such as
   * nodes loaded from a lockfile that lacks a loaded manifest.
   * This field should be used to retrieve package versions instead.
   */
  version?: string

  /**
   * An address {@link PackageInfoClient} may use to extract this package.
   */
  resolved?: string

  /**
   * The file system location for this node.
   */
  get location(): string {
    if (this.#location) {
      return this.#location
    }
    this.#location = `./node_modules/.vlt/${this.id}/node_modules/${this.name}`
    // if using the default location, it is in the store
    this.inVltStore = () => true
    return this.#location
  }

  set location(location: string) {
    this.#location = location
    // reset memoization, since it might be elsewhere now
    if (this.inVltStore !== Node.prototype.inVltStore) {
      this.inVltStore = Node.prototype.inVltStore
    }
  }

  constructor(
    options: NodeOptions,
    id?: DepID,
    manifest?: ManifestMinified,
    spec?: Spec,
    name?: string,
    version?: string,
  ) {
    this.#options = options
    this.projectRoot = options.projectRoot
    if (id) {
      this.id = id
    } else {
      if (!manifest || !spec) {
        throw typeError(
          'A new Node needs either a manifest & spec or an id parameter',
          {
            manifest,
          },
        )
      }
      this.id = getId(spec, manifest)
    }
    this.graph = options.graph
    this.manifest = manifest
    this.#name = name || this.manifest?.name
    this.version = version || this.manifest?.version
  }

  /**
   * return true if this node is located in the vlt store
   * memoized the first time it's called, since the store location
   * doesn't change within the context of a single operation.
   */
  inVltStore(): boolean {
    // technically this just means it's in *a* vlt store, but we can safely
    // assume that a user won't construct a path like this by accident,
    // and there's only ever one store in any given project.
    const inStore = this.location.endsWith(
      `.vlt/${this.id}/node_modules/${this.name}`,
    )
    this.inVltStore = () => inStore
    return inStore
  }

  #registryNodeResolved(tuple: DepIDTuple) {
    const spec = hydrateTuple(tuple, this.#name, this.#options)
    this.resolved =
      this.manifest?.dist?.tarball || spec.conventionalRegistryTarball
    this.integrity ??= this.manifest?.dist?.integrity
  }

  equals(other: Node) {
    return this.id === other.id && this.location === other.location
  }

  /**
   * Sets the node as an importer along with its location.
   */
  setImporterLocation(location: string) {
    this.#location = location
    this.importer = true
  }

  /**
   * Sets the appropriate resolve / integrity value for this node.
   * Note that other places might also set these values, like for
   * example the lockfile that might have already have this info.
   */
  setResolved() {
    // file | remote | workspace type of ids all points to a URI that
    // can be used as the `resolved` value, so we split the dep id
    // for these cases.
    const tuple = splitDepID(this.id)
    const [type, resolved] = tuple
    switch (type) {
      case 'registry':
        this.#registryNodeResolved(tuple)
        break
      default:
        this.resolved = resolved
        break
    }
  }

  setDefaultLocation() {
    const def = `./node_modules/.vlt/${this.id}/node_modules/${this.name}`

    // only relocate if the location is in node_modules already
    if (
      !this.importer &&
      (!this.#location ||
        (this.#location !== def &&
          /^(?:\.\/)?node_modules\//.test(this.#location)))
    ) {
      this.#location = def
    }
  }

  /**
   * Add an edge from this node connecting it to a direct dependency.
   */
  addEdgesTo(type: DependencyTypeShort, spec: Spec, node?: Node) {
    const edge = new Edge(type, spec, this, node)
    node?.edgesIn.add(edge)
    this.edgesOut.set(spec.name, edge)
    return edge
  }

  toJSON() {
    return {
      id: this.id,
      name: this.name,
      version: this.version,
      location: this.location,
      importer: this.importer,
      manifest: this.manifest,
      projectRoot: this.projectRoot,
      integrity: this.integrity,
      resolved: this.resolved,
      dev: this.dev,
      optional: this.optional,
    }
  }

  toString() {
    const version = this.version ? `@${this.version}` : ''
    const depIdTuple = splitDepID(this.id)
    switch (depIdTuple[0]) {
      case 'registry': {
        const prefix = depIdTuple[1] ? `${depIdTuple[1]}:` : ''
        return `${prefix}${depIdTuple[2]}`
      }
      case 'git':
      case 'workspace':
      case 'file':
      case 'remote': {
        const nameVersion =
          this.name !== this.id ? `:${this.name}${version}` : ''
        return `${depIdTuple[0]}(${depIdTuple[1]})${nameVersion}`
      }
    }
    /* c8 ignore next - impossible */
  }
}
