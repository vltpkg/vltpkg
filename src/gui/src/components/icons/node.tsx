import { createLucideIcon } from 'lucide-react'

export const Node = createLucideIcon('Node', [
  [
    'path',
    {
      d: 'M11.8751 22C11.5993 22 11.3423 21.9266 11.1032 21.7977L8.65943 20.3461C8.29224 20.1437 8.47584 20.0703 8.58599 20.0336C9.08209 19.8679 9.17428 19.8312 9.68835 19.5375C9.74382 19.5008 9.81725 19.5187 9.87194 19.5562L11.7462 20.6773C11.8196 20.7141 11.9118 20.7141 11.9665 20.6773L19.2978 16.4328C19.3713 16.396 19.408 16.3226 19.408 16.2304V7.76002C19.408 7.66783 19.3713 7.59439 19.2978 7.55767L11.9665 3.33185C11.8931 3.29513 11.8009 3.29513 11.7462 3.33185L4.41486 7.55767C4.34142 7.59439 4.3047 7.68658 4.3047 7.76002V16.2304C4.3047 16.3039 4.34142 16.396 4.41486 16.4328L6.418 17.5906C7.50239 18.1422 8.18208 17.4984 8.18208 16.8554V8.49518C8.18208 8.38502 8.27427 8.27487 8.4024 8.27487H9.33912C9.44928 8.27487 9.55944 8.36705 9.55944 8.49518V16.8554C9.55944 18.307 8.76959 19.1523 7.39145 19.1523C6.96879 19.1523 6.63832 19.1523 5.70081 18.6929L3.77188 17.5906C3.29453 17.3148 3 16.8007 3 16.2492V7.77877C3 7.2272 3.29375 6.71313 3.77188 6.43735L11.1032 2.19278C11.5626 1.93574 12.1876 1.93574 12.647 2.19278L19.9783 6.43735C20.4556 6.71313 20.7502 7.2272 20.7502 7.77877V16.2492C20.7502 16.8007 20.4564 17.3148 19.9783 17.5906L12.647 21.8352C12.4079 21.9445 12.1321 22 11.8751 22ZM14.1353 16.1757C10.9196 16.1757 10.2587 14.7062 10.2587 13.4562C10.2587 13.346 10.3509 13.2359 10.479 13.2359H11.4345C11.5446 13.2359 11.6368 13.3093 11.6368 13.4194C11.7837 14.3929 12.2063 14.871 14.154 14.871C15.6978 14.871 16.3587 14.5218 16.3587 13.6952C16.3587 13.2179 16.1751 12.8687 13.7681 12.6296C11.7649 12.4273 10.5157 11.9866 10.5157 10.3882C10.5157 8.89987 11.7649 8.01783 13.8595 8.01783C16.2111 8.01783 17.3689 8.82643 17.5158 10.5905C17.5158 10.646 17.497 10.7007 17.4603 10.7561C17.4236 10.7929 17.3681 10.8296 17.3134 10.8296H16.3579C16.2658 10.8296 16.1743 10.7561 16.1556 10.664C15.9353 9.653 15.3657 9.32253 13.8587 9.32253C12.1681 9.32253 11.9665 9.91082 11.9665 10.3514C11.9665 10.8843 12.2056 11.0499 14.4837 11.3436C16.7439 11.6374 17.8095 12.0601 17.8095 13.6405C17.7916 15.257 16.4689 16.1757 14.1353 16.1757Z',
      key: '1',
      fill: 'currentColor',
      strokeWidth: '0',
    },
  ],
])

export const NodeWordmark = createLucideIcon('NodeWordmark', [
  [
    'path',
    {
      d: 'M21.8928 12.5235C21.9591 12.4848 22 12.414 22 12.3371V11.8385C22 11.7618 21.9593 11.6911 21.8928 11.6523L20.1212 10.6237C20.0545 10.585 19.9723 10.5848 19.9053 10.6233L18.1227 11.6525C18.0562 11.6911 18.0151 11.762 18.0151 11.8389V13.8969C18.0151 13.9743 18.0566 14.0459 18.124 14.0842L19.8954 15.0935C19.9606 15.131 20.0407 15.1312 20.1066 15.0948L21.1779 14.4991C21.212 14.4801 21.2332 14.4445 21.2334 14.4059C21.2336 14.367 21.2129 14.331 21.1792 14.3117L19.3855 13.2823C19.3519 13.263 19.3313 13.2274 19.3313 13.1891V12.5439C19.3313 12.5053 19.3519 12.4699 19.3851 12.4509L19.9437 12.1289C19.9767 12.1096 20.0178 12.1096 20.0511 12.1289L20.6096 12.4509C20.6431 12.4701 20.6636 12.5055 20.6636 12.5439V13.0514C20.6636 13.0901 20.6842 13.1257 20.7174 13.1448C20.7509 13.1641 20.792 13.1641 20.8255 13.1446L21.8928 12.5235ZM19.9869 12.4269C19.9998 12.4197 20.0155 12.4197 20.0282 12.4269L20.3703 12.6246C20.383 12.632 20.3908 12.6457 20.3908 12.6604V13.0554C20.3908 13.0702 20.383 13.0838 20.3703 13.0912L20.0282 13.2886C20.0155 13.296 19.9996 13.296 19.9869 13.2886L19.645 13.0912C19.6321 13.0838 19.6243 13.0702 19.6243 13.0554V12.6604C19.6243 12.6457 19.6321 12.632 19.645 12.6246L19.9869 12.4269ZM15.3479 13.2488C15.3479 13.2678 15.3377 13.2856 15.321 13.2954L14.7088 13.6483C14.6921 13.658 14.6715 13.658 14.655 13.6483L14.0429 13.2954C14.0261 13.2856 14.016 13.2678 14.016 13.2488V12.5417C14.016 12.5227 14.0264 12.5047 14.0429 12.4949L14.655 12.1416C14.6718 12.1317 14.6923 12.1317 14.709 12.1416L15.3214 12.4949C15.3379 12.5047 15.3481 12.5227 15.3481 12.5417L15.3479 13.2488ZM15.5133 8.01351C15.4801 7.99508 15.4394 7.99551 15.4063 8.01478C15.3735 8.03427 15.3532 8.06943 15.3532 8.10756V10.8432C15.3532 10.8699 15.3388 10.8949 15.3157 10.9084C15.2924 10.9218 15.2636 10.9218 15.2403 10.9084L14.7938 10.6511C14.7273 10.6127 14.6453 10.6127 14.5788 10.6511L12.7957 11.6801C12.729 11.7186 12.6879 11.7898 12.6879 11.8667V13.9253C12.6879 14.0022 12.729 14.0732 12.7955 14.1119L14.5786 15.1416C14.6451 15.1801 14.7273 15.1801 14.794 15.1416L16.5775 14.1119C16.644 14.0732 16.6851 14.0022 16.6851 13.9253V8.79363C16.6851 8.71568 16.6427 8.64367 16.5747 8.60554L15.5133 8.01351ZM9.23801 10.6123C9.30452 10.5737 9.3867 10.5737 9.45322 10.6123L11.2363 11.6411C11.3028 11.6796 11.3439 11.7508 11.3439 11.8277V13.887C11.3439 13.9639 11.3028 14.0351 11.2363 14.0736L9.45343 15.1032C9.38692 15.1416 9.30473 15.1416 9.23801 15.1032L7.45557 14.0736C7.38906 14.0351 7.34796 13.9639 7.34796 13.887V11.8275C7.34796 11.7508 7.38906 11.6796 7.45557 11.6411L9.23801 10.6123ZM5.99551 11.8105C5.99551 11.733 5.95484 11.6619 5.88791 11.6235L4.10546 10.5977C4.07496 10.5799 4.04128 10.5708 4.00718 10.5695H3.98875C3.95486 10.5708 3.92097 10.5799 3.89047 10.5977L2.10782 11.6235C2.04152 11.6619 2 11.733 2 11.8105L2.00381 14.5726C2.00381 14.6112 2.02394 14.647 2.05761 14.6656C2.09066 14.6855 2.13154 14.6855 2.16458 14.6656L3.22367 14.059C3.2906 14.0192 3.33148 13.9491 3.33148 13.8722V12.5818C3.33148 12.5047 3.37215 12.4337 3.43887 12.3952L3.89005 12.1355C3.92351 12.1158 3.96037 12.1065 3.99765 12.1065C4.0345 12.1065 4.07221 12.1158 4.10504 12.1355L4.55579 12.3952C4.62251 12.4337 4.6636 12.5047 4.6636 12.5818V13.8722C4.6636 13.9491 4.70491 14.0196 4.77142 14.059L5.8305 14.6656C5.86376 14.6855 5.90528 14.6855 5.93811 14.6656C5.97094 14.6468 5.99149 14.611 5.99149 14.5726L5.99551 11.8105Z',
      key: '1',
      fill: 'currentColor',
      strokeWidth: '0',
    },
  ],
])
