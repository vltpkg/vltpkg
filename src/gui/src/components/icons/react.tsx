export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
