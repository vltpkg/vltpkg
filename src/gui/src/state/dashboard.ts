import { create } from 'zustand'
import type { Table, VisibilityState } from '@tanstack/react-table'
import type { DashboardDataProject } from '@/state/types.ts'

export type DashboardState = {
  currentView: 'grid' | 'table'
  table: Table<DashboardDataProject> | undefined
  tableFilterValue: string
  filteredProjects: DashboardDataProject[]
  columnVisibility: VisibilityState
}

export type DashboardAction = {
  updateCurrentView: (
    currentView: DashboardState['currentView'],
  ) => void
  updateTable: (table: DashboardState['table']) => void
  setTableFilterValue: (
    tableFilterValue: DashboardState['tableFilterValue'],
  ) => void
  setFilteredProjects: (
    filteredProjects: DashboardDataProject[],
  ) => void
  setColumnVisibility: (
    columnVisibility: DashboardState['columnVisibility'],
  ) => void
}

const initialState: DashboardState = {
  currentView: 'grid',
  table: undefined,
  tableFilterValue: '',
  filteredProjects: [],
  columnVisibility: {
    type: false,
    private: false,
    version: false,
  },
}

export const useDashboardStore = create<
  DashboardState & DashboardAction
>(set => {
  const store = {
    ...initialState,
    updateCurrentView: (currentView: DashboardState['currentView']) =>
      set(() => ({ currentView })),
    updateTable: (table: DashboardState['table']) =>
      set(() => ({ table })),
    setTableFilterValue: (
      tableFilterValue: DashboardState['tableFilterValue'],
    ) => set(() => ({ tableFilterValue })),
    setFilteredProjects: (filteredProjects: DashboardDataProject[]) =>
      set(() => ({ filteredProjects })),
    setColumnVisibility: (
      columnVisibility: DashboardState['columnVisibility'],
    ) => set(() => ({ columnVisibility })),
  }
  return store
})
