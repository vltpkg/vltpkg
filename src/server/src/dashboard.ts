import { getUser } from '@vltpkg/git'
import { getAuthorFromGitUser } from '@vltpkg/init'
import type { PackageJson } from '@vltpkg/package-json'
import type { Manifest } from '@vltpkg/types'
import { writeFileSync } from 'node:fs'
import { rm } from 'node:fs/promises'
import { homedir } from 'node:os'
import { resolve } from 'node:path'
import type { PathBase, PathScurry } from 'path-scurry'
import { getReadablePath } from './get-readable-path.ts'
import type { ProjectTools } from './project-tools.ts'
import { inferTools } from './project-tools.ts'
import { readProjectFolders } from './read-project-folders.ts'

// TODO: rather than writing a dashboard.json file, keep the dashboard
// info in memory, and make a handle to the Dashboard object persist in
// the VltServer class. Then, if we get a request for /dashboard.json,
// just server it from the dashboard object without reading static file.

export type DashboardProjectData = {
  name: string
  readablePath: string
  path: string
  manifest: Manifest
  tools: ProjectTools[]
  mtime?: number
}

export type DashboardOptions = {
  'dashboard-root': string[]
  scurry: PathScurry
  packageJson: PackageJson
  publicDir: string
}

export type DashboardLocation = {
  path: string
  readablePath: string
}

export type DashboardData = {
  cwd: string
  dashboardProjectLocations: DashboardLocation[]
  defaultAuthor: string
  projects: DashboardProjectData[]
}

/**
 * Class to handle updating and formatting dashboard data
 */
export class Dashboard {
  packageJson: PackageJson
  dashboardRoot: string[]
  scurry: PathScurry
  publicDir: string

  constructor(options: DashboardOptions) {
    const {
      'dashboard-root': dashboardRoot,
      scurry,
      packageJson,
      publicDir,
    } = options
    this.packageJson = packageJson
    this.dashboardRoot = dashboardRoot
    if (!this.dashboardRoot.length) this.dashboardRoot = [homedir()]
    this.scurry = scurry
    this.publicDir = publicDir
  }

  async update() {
    // TODO: projectFolders should be set once, not every time
    // Store the mtime for every folder that we find, and only update
    // the ones that are actually modified.
    const dashboard = await this.format(
      await readProjectFolders({
        scurry: this.scurry,
        userDefinedProjectPaths: this.dashboardRoot,
      }),
    )
    const hasDashboard = dashboard.projects.length > 0

    // only need to write the file if we're not about to delete it.
    if (!hasDashboard) {
      await rm(resolve(this.publicDir, 'dashboard.json'), {
        force: true,
      })
    } else {
      const dashboardJson = JSON.stringify(dashboard, null, 2)
      writeFileSync(
        resolve(this.publicDir, 'dashboard.json'),
        dashboardJson,
      )
    }

    return hasDashboard
  }

  async format(projectFolders: PathBase[]): Promise<DashboardData> {
    const userDefinedProjectPaths = this.dashboardRoot.map(
      path =>
        ({
          path,
          readablePath: getReadablePath(path),
        }) as DashboardLocation,
    )

    const result: DashboardData = {
      cwd: process.cwd(),
      dashboardProjectLocations: projectFolders
        .map((dir: PathBase) => {
          const path = dir.resolve('../').fullpath()
          const res: DashboardLocation = {
            path,
            readablePath: getReadablePath(path),
          }
          return res
        })
        .concat(userDefinedProjectPaths)
        .reduce<DashboardLocation[]>((acc, curr) => {
          if (acc.every(obj => obj.path !== curr.path)) {
            acc.push(curr)
          }
          return acc
        }, [])
        .sort(
          (a, b) => a.readablePath.length - b.readablePath.length,
        ),
      defaultAuthor: getAuthorFromGitUser(
        await getUser().catch(() => undefined),
      ),
      projects: [],
    }

    for (const folder of projectFolders) {
      const projectData = this.getProjectData(folder)
      if (projectData) {
        result.projects.push(projectData)
      }
    }

    return result
  }

  getProjectData(folder: PathBase): DashboardProjectData | undefined {
    let manifest
    try {
      manifest = this.packageJson.read(folder.fullpath())
    } catch {
      return
    }
    const path = folder.fullpath()
    return {
      name: manifest.name || folder.name,
      readablePath: getReadablePath(path),
      path,
      manifest,
      tools: inferTools(manifest, folder, this.scurry),
      mtime: folder.lstatSync()?.mtimeMs,
    }
  }
}
