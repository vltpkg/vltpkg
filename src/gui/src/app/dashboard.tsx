import { useEffect } from 'react'
import { DashboardGrid } from '@/components/dashboard-grid/index.jsx'
import { useGraphStore } from '@/state/index.js'
import { startDashboardData } from '@/lib/start-dashboard-data.js'

export const Dashboard = () => {
  const dashboard = useGraphStore(state => state.dashboard)
  const updateActiveRoute = useGraphStore(
    state => state.updateActiveRoute,
  )
  const updateDashboard = useGraphStore(
    state => state.updateDashboard,
  )
  const updateErrorCause = useGraphStore(
    state => state.updateErrorCause,
  )
  const stamp = useGraphStore(state => state.stamp)

  useEffect(() => {
    startDashboardData({
      updateActiveRoute,
      updateErrorCause,
      updateDashboard,
      stamp,
    })

    history.pushState(
      { query: '', route: '/dashboard' },
      '',
      '/dashboard',
    )
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    window.scrollTo?.(0, 0)
  }, [stamp])

  return (
    <section className="flex min-h-[80svh] w-full grow flex-col bg-white dark:bg-black">
      <div className="flex w-full items-center justify-between border-b-[1px] border-t-[1px] border-solid px-8 py-4">
        {dashboard?.cwd ?
          <p className="font-mono text-xs font-light text-muted-foreground">
            Directory: {dashboard.cwd}
          </p>
        : ''}
        {dashboard?.buildVersion ?
          <p className="text-right font-mono text-xs font-light text-muted-foreground">
            build: v{dashboard.buildVersion}
          </p>
        : ''}
      </div>
      <DashboardContent />
    </section>
  )
}

const DashboardContent = () => {
  return <DashboardGrid />
}
