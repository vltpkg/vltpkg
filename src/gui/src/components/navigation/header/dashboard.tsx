import { motion } from 'framer-motion'
import { SortDropdown } from '@/components/sort-dropdown.tsx'
import { Button } from '@/components/ui/button.tsx'
import { Plus } from 'lucide-react'
import { FilterSearch } from '@/components/ui/filter-search.tsx'
import { TableFilterSearch } from '@/components/data-table/table-filter-search.tsx'
import { TableViewDropdown } from '@/components/data-table/table-view-dropdown.tsx'
import { DashboardViewToggle } from '@/components/dashboard-grid/dashboard-view-toggle.tsx'
import { NavLink } from 'react-router'
import { useDashboardStore } from '@/state/dashboard.ts'
import { useGraphStore } from '@/state/index.ts'

export const DashboardHeader = () => {
  const currentView = useDashboardStore(state => state.currentView)
  const setCurrentView = useDashboardStore(
    state => state.updateCurrentView,
  )
  const table = useDashboardStore(state => state.table)
  const tableFilterValue = useDashboardStore(
    state => state.tableFilterValue,
  )
  const setTableFilterValue = useDashboardStore(
    state => state.setTableFilterValue,
  )
  const dashboard = useGraphStore(state => state.dashboard)
  const filteredProjects = useDashboardStore(
    state => state.filteredProjects,
  )
  const setFilteredProjects = useDashboardStore(
    state => state.setFilteredProjects,
  )
  const columnVisibility = useDashboardStore(
    state => state.columnVisibility,
  )
  const setColumnVisibility = useDashboardStore(
    state => state.setColumnVisibility,
  )

  return (
    <div className="flex w-full items-center gap-5">
      {currentView === 'table' ?
        <TableFilterSearch
          filterValue={tableFilterValue}
          onFilterChange={setTableFilterValue}
          className="w-full"
        />
      : <FilterSearch
          placeholder="Filter Projects"
          items={dashboard?.projects ?? []}
          setFilteredItems={setFilteredProjects}
          className="w-full"
        />
      }
      <DashboardViewToggle
        currentView={currentView}
        setCurrentView={setCurrentView}
        classNames={{
          toggleClassName: 'rounded-xl',
          sliderClassName: 'rounded-[9px]',
          optionClassName: 'rounded-lg',
        }}
      />
      {currentView === 'table' ?
        <motion.div
          key={currentView}
          initial={{ opacity: 0 }}
          exit={{ opacity: 0 }}
          animate={{ opacity: 1 }}>
          <TableViewDropdown
            columnVisibility={columnVisibility}
            setColumnVisibility={setColumnVisibility}
            table={table}
            className="w-[100px] rounded-xl md:w-[120px]"
          />
        </motion.div>
      : <motion.div
          key={currentView}
          exit={{ opacity: 0 }}
          initial={{ opacity: 1 }}
          animate={{ opacity: 1 }}>
          <SortDropdown
            items={filteredProjects}
            className="w-[100px] rounded-xl md:w-[120px]"
            setFilteredItems={value => {
              const newValue =
                typeof value === 'function' ?
                  value(filteredProjects)
                : value
              setFilteredProjects(newValue)
            }}
            sortKey="name"
          />
        </motion.div>
      }
      <Button asChild className="ml-auto rounded-xl">
        <NavLink to="/create-new-project">
          <Plus size={24} />
          <span className="hidden md:flex">Create New Project</span>
        </NavLink>
      </Button>
    </div>
  )
}
