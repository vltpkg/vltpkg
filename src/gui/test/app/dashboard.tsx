import {
  test,
  expect,
  vi,
  afterAll,
  afterEach,
  beforeAll,
} from 'vitest'
import { setupServer } from 'msw/node'
import { http, HttpResponse } from 'msw'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.js'
import { Dashboard } from '@/app/dashboard.jsx'

vi.mock('@/components/ui/logo.jsx', () => ({
  Logo: 'gui-logo',
}))
vi.mock('@/components/ui/title.jsx', () => ({
  Title: 'gui-title',
}))
vi.mock('@/components/ui/theme-switcher.jsx', () => ({
  ThemeSwitcher: 'gui-theme-switcher',
}))
vi.mock('@/components/dashboard-grid/index.jsx', () => ({
  DashboardGrid: 'gui-dashboard-grid',
}))

export const restHandlers = [
  http.get('/dashboard.json', () => {
    return HttpResponse.json([
      {
        name: 'project-foo',
        path: '/home/user/project-foo',
        manifest: { name: 'project-foo', version: '1.0.0' },
        tools: ['node', 'vlt'],
        mtime: 1730498483044,
      },
      {
        name: 'project-bar',
        path: '/home/user/project-bar',
        manifest: { name: 'project-bar', version: '1.0.0' },
        tools: ['pnpm'],
        mtime: 1730498491029,
      },
    ])
  }),
]

const server = setupServer(...restHandlers)

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

beforeAll(() => server.listen())

afterEach(() => {
  server.resetHandlers()
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

afterAll(() => server.close())

test('render default', async () => {
  render(<Dashboard />)
  expect(window.document.body.innerHTML).toMatchSnapshot()
})
