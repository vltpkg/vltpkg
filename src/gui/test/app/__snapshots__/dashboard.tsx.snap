// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render default 1`] = `

<div>
  <section class="flex h-full max-h-[calc(100svh-65px-16px)] w-full grow flex-col rounded-b-lg border-x-[1px] border-b-[1px]">
    <gui-dashboard-grid>
    </gui-dashboard-grid>
  </section>
</div>

`;
